src/utils/transformer-utils.ts

```diff
@@ -1,188 +1,287 @@
+} from "../transformers/transformer-models.js";
 
-export const createSpan = (
-    guid: string,
-    marks?: string[],
-    text?: string
-): IPortableTextSpan => {
-    return {
-        _type: 'span',
-        _key: guid,
-        marks: marks || [],
-        text: text || ''
+export const textStyleElements = ['strong', 'em', 'sub', 'sup', 'code'] as const;
+export const blockElements = ['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;
+export const ignoredElements = ['img', 'tbody', 'ol', 'ul'] as const;
+export const tableElements = ['table', 'td', 'tr'] as const;
+export const lineBreakElement = 'br' as const;
+export const anchorElement = 'a' as const;
+export const objectElement = 'object' as const;
+export const assetElement = 'figure' as const;
+export const listItemElement = 'li' as const;
+export const markElements = [...textStyleElements, anchorElement] as const;
+export const allElements = [
+    ...blockElements,
+    ...ignoredElements,
+    ...markElements,
+    ...tableElements,
+    assetElement,
+    objectElement,
+    lineBreakElement,
+    listItemElement,
+  ] as const;
+
+export type TextStyleElement = typeof textStyleElements[number];
+export type BlockElement = typeof blockElements[number];
+export type IgnoredElement = typeof ignoredElements[number];
+export type MarkElement = typeof markElements[number];
+export type ValidElement = typeof allElements[number];
+
+export type TransformLinkFunction = (node: IDomHtmlNode) => [PortableTextLink, PortableTextMark];
+export type TransformElementFunction = (node: IDomHtmlNode) => PortableTextObject[];
+export type TransformListItemFunction = (node: IDomHtmlNode, depth: number, listType: PortableTextListItemType) => PortableTextStrictListItemBlock[];
+export type TransformTextFunction = (node: IDomTextNode) => PortableTextSpan;
+export type TransformTableCellFunction = (node: IDomHtmlNode) => PortableTextObject[];
+export type TransformFunction = TransformElementFunction | TransformListItemFunction;
+
+export type MergePortableTextItemsFunction = (itemsToMerge: ReadonlyArray<PortableTextObject>) => PortableTextObject[];
+
+/**
+ * Recursively traverses and optionally transforms a Portable Text structure using a provided 
+ * callback function. The callback is applied to each node in the structure. If the callback 
+ * does not modify a node, the original node is used.
+ *
+ * @template T The type of the Portable Text nodes, defaulting to PortableTextObject.
+ * @param {T} object - The root node of the Portable Text structure to be traversed. 
+ *   It can be a default Portable Text object or a custom type that extends from it.
+ * @param {(object: T) => ArbitraryTypedObject | undefined} callback - A callback function 
+ *   invoked for each node in the Portable Text structure. It can return a modified version 
+ *   of the node or `undefined` if no modifications are to be made.
+ * @returns {ArbitraryTypedObject} - A modified copy of the original portable text structure.
```

Comment:
You don't copy the original structure, you modify it (unless the callback call copies it). So this should be explicitly mentioned (that the original object is mutated).

Reply:
modified to create a copy with spread operator, adjusted logic to not traverse "marks" array